nixos/modules/services/web-servers/hitch/default.nix

```diff
@@ -0,0 +1,109 @@
+{ config, lib, pkgs, ...}:
+let
+  cfg = config.services.hitch;
+  hitchConfig = with lib; pkgs.writeText "hitch.conf" (concatStringsSep "\n" [
+    ("backend = \"${cfg.backend}\"")
+    ("frontend = \"${cfg.frontend}\"")
+    (concatMapStrings (s: "pem-file = \"${s}\"\n") cfg.pem-files)
+    ("ciphers = \"${cfg.ciphers}\"")
+    ("ocsp-dir = \"${cfg.ocspStaplingDir}\"")
+    "user = \"${cfg.user}\""
+    "group = \"${cfg.group}\""
+    cfg.extraConfig
+  ]);
+in
+with lib;
+{
+  options = {
+    services.hitch = {
+      enable = mkEnableOption "Hitch Server";
+
+      backend = mkOption {
+        type = types.str;
+        description = ''
+          The host and port Hitch connects to when receiving
+          a connection
+        '';
+      };
+
+      ciphers = mkOption {
+        type = types.str;
+        default = "EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH";
+        description = "The list of ciphers to use";
+      };
+
+      frontend = mkOption {
+        type = types.either types.str (types.listOf types.str);
+        default = "[127.0.0.1]:443";
+        description = ''
+          This specifies the port and interface (the listen endpoint) that Hitch
+          binds to when listening for connections. In the form [HOST]:PORT[+CERT]
+        '';
+      };
+
+      pem-files = mkOption {
+        type = types.listOf types.path;
+        default = [];
+        description = "PEM files to use";
+      };
+
+      ocspStaplingDir = mkOption {
+        type = types.path;
+        default = "/var/run/hitch/ocsp-cache";
+        description = "The location of the OCSP Stapling cache";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "hitch";
+        description = "The user to run as";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "hitch";
+        description = "The group to run as";
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = "Additional configuration lines";
+      };
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.services.hitch = {
+      description = "Hitch";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      preStart = ''
+        ${pkgs.hitch}/sbin/hitch -t --config ${hitchConfig}
+        mkdir -p ${cfg.ocspStaplingDir}
+        chown -R hitch:hitch ${cfg.ocspStaplingDir}
+      '';
+      postStop = ''
+        rm -rf ${cfg.ocspStaplingDir}
+      '';
```

Comment:
I simplified the module a bit by making use of systemd's RuntimeDirectory.
What do you think?

```diff
diff --git a/nixos/modules/services/web-servers/hitch/default.nix b/nixos/modules/services/web-servers/hitch/default.nix
index 54663082be1..87ee69aa7cd 100644
--- a/nixos/modules/services/web-servers/hitch/default.nix
+++ b/nixos/modules/services/web-servers/hitch/default.nix
@@ -6,7 +6,7 @@ let
     ("frontend = \"${cfg.frontend}\"")
     (concatMapStrings (s: "pem-file = \"${s}\"\n") cfg.pem-files)
     ("ciphers = \"${cfg.ciphers}\"")
-    ("ocsp-dir = \"${cfg.ocspStaplingDir}\"")
+    ("ocsp-dir = \"/run/hitch/ocsp-cache\"")
     "user = \"${cfg.user}\""
     "group = \"${cfg.group}\""
     cfg.extraConfig
@@ -22,7 +22,7 @@ with lib;
         type = types.str;
         description = ''
           The host and port Hitch connects to when receiving
-          a connection
+          a connection in the form [HOST]:PORT
         '';
       };
 
@@ -36,8 +36,8 @@ with lib;
         type = types.either types.str (types.listOf types.str);
         default = "[127.0.0.1]:443";
         description = ''
-          This specifies the port and interface (the listen endpoint) that Hitch
-          binds to when listening for connections. In the form [HOST]:PORT[+CERT]
+          This specifies the port and interface of the listen endpoint in the
+          form [HOST]:PORT[+CERT].
         '';
       };
 
@@ -47,12 +47,6 @@ with lib;
         description = "PEM files to use";
       };
 
-      ocspStaplingDir = mkOption {
-        type = types.path;
-        default = "/var/run/hitch/ocsp-cache";
-        description = "The location of the OCSP Stapling cache";
-      };
-
       user = mkOption {
         type = types.str;
         default = "hitch";
@@ -80,16 +74,10 @@ with lib;
       description = "Hitch";
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
-      preStart = ''
-        ${pkgs.hitch}/sbin/hitch -t --config ${hitchConfig}
-        mkdir -p ${cfg.ocspStaplingDir}
-        chown -R hitch:hitch ${cfg.ocspStaplingDir}
-      '';
-      postStop = ''
-        rm -rf ${cfg.ocspStaplingDir}
-      '';
       serviceConfig = {
+        RuntimeDirectory = [ "hitch/ocsp-cache" ];
         Type = "forking";
+        ExecStartPre = "${pkgs.hitch}/sbin/hitch -t --config ${hitchConfig}";
         ExecStart = "${pkgs.hitch}/sbin/hitch --daemon --config ${hitchConfig}";
         ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
         Restart = "always";
@@ -100,10 +88,7 @@ with lib;
 
     environment.systemPackages = [ pkgs.hitch ];
 
-    users.extraUsers.hitch = {
-      group = "hitch";
-    };
-
+    users.extraUsers.hitch.group = "hitch";
     users.extraGroups.hitch = {};
   };
 }
```

Why do you delete this directory here? This is different from what upstream does

Reply:
Hmm. We can do that, but then the runtime directory will be owned by root in this case, no?

Reply:
If we go that route, then we'd want to specify systemd's User and Group. I'm not sure how hitch handles being run as non-root. I'll try it out.

Reply:
`RuntimeDirectory = [ "hitch" "hitch/ocsp-cache" ];` maybe then?

Reply:
Ah. you are right, usually the service is also started as non-root in this case.

Reply:
In theory this could be circumvented by starting the service as the right user and give it ambient capability to bind privileged ports: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/traefik.nix#L104

Reply:
This might have a security impact because before hitch would not be able to bind new ports after root privileges are dropped. At least you could remove the `rm` command and update the two option description I modified.

Reply:
Done. I actually like the RuntimeDirectory/CacheDirectory route but I suppose we can start with this.